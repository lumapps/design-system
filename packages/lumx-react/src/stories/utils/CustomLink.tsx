/* eslint-disable jsx-a11y/anchor-has-content,jsx-a11y/anchor-is-valid */
import React from 'react';

/**
 * Example custom link that can be used in linkAs props.
 */
export const CustomLink: any = ({ ...props }) => <a href="#" data-custom-link {...props} />;
