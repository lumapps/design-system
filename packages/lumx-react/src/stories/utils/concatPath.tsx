export type PathPart = string | number;
export type Path = PathPart | Array<PathPart>;

/**
 * Concat flatten object path
 *
 * @example concatPath('foo', 'bar') // => 'foo.bar'
 * @example concatPath(['foo', 0]) // => 'foo[0]'
 * @example concatPath('foo', 0, ['bar']) // => 'foo[0].bar'
 */
export const concatPath = (...prefix: Path[]) => {
    const [first, ...rest] = prefix.flat();
    return rest.reduce<string>((acc, part) => {
        if (typeof part === 'number') return `${acc}[${part}]`;
        return `${acc}.${part}`;
    }, String(first));
};
