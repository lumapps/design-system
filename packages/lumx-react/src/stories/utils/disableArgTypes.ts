export function disableArgTypes(argTypes: string[]) {
    return Object.fromEntries(argTypes.map((argType) => [argType, { table: { disable: true } }]));
}
