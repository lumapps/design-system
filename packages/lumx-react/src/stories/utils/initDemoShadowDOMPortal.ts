/** Init a demo shadow DOM to use as portal container */
export const initDemoShadowDOMPortal = () => {
    const div = document.createElement('div');
    document.body.appendChild(div);
    const container = div.attachShadow({ mode: 'open' });
    const style = document.createElement('style');
    style.innerText = `button { color: red; }`;
    container.appendChild(style);
    return { container, teardown: () => div.remove() };
};
