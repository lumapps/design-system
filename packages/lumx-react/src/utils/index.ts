/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
