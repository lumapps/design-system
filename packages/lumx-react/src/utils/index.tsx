export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
