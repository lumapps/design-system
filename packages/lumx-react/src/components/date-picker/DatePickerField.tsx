import { DatePicker, Placement, Popover, TextField, IconButtonProps } from '@lumx/react';
import { useFocusTrap } from '@lumx/react/hooks/useFocusTrap';

import moment from 'moment';

import React, { forwardRef, SyntheticEvent, useCallback, useRef, useState } from 'react';

import { useFocus } from '@lumx/react/hooks/useFocus';
import { Comp, GenericProps } from '@lumx/react/utils';

/**
 * Defines the props of the component.
 */
export interface DatePickerFieldProps extends GenericProps {
    /** Default month. */
    defaultMonth?: Date;
    /** Whether the component is disabled or not. */
    isDisabled?: boolean;
    /** Locale (language or region) to use. */
    locale: string;
    /** Date after which dates can't be selected. */
    maxDate?: Date;
    /** Date before which dates can't be selected. */
    minDate?: Date;
    /** Native input name property. */
    name?: string;
    /** Props to pass to the next month button (minus those already set by the DatePickerControlled props). */
    nextButtonProps: Pick<IconButtonProps, 'label'> & Omit<IconButtonProps, 'label' | 'onClick' | 'icon' | 'emphasis'>;
    /** Props to pass to the previous month button (minus those already set by the DatePickerControlled props). */
    previousButtonProps: Pick<IconButtonProps, 'label'> &
        Omit<IconButtonProps, 'label' | 'onClick' | 'icon' | 'emphasis'>;
    /** Currently selected date. */
    value: Date | undefined;
    /** On change callback. */
    onChange(value: Date | undefined, name?: string, event?: SyntheticEvent): void;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'DatePickerField';

/**
 * DatePickerField component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const DatePickerField: Comp<DatePickerFieldProps, HTMLDivElement> = forwardRef((props, ref) => {
    const {
        defaultMonth,
        disabled,
        isDisabled = disabled,
        locale,
        maxDate,
        minDate,
        name,
        nextButtonProps,
        onChange,
        previousButtonProps,
        value,
        ...forwardedProps
    } = props;
    const wrapperRef = useRef(null);
    const anchorRef = useRef(null);

    const [isOpen, setIsOpen] = useState(false);

    const toggleSimpleMenu = () => {
        setIsOpen(!isOpen);
    };

    const onClose = useCallback(() => {
        setIsOpen(false);
    }, []);

    useFocus(anchorRef.current, isOpen);
    const handleKeyboardNav = (evt: React.KeyboardEvent) => {
        if ((evt.key === 'Enter' || evt.key === ' ') && toggleSimpleMenu) {
            toggleSimpleMenu();
        }
    };

    // Handle focus trap.
    const todayOrSelectedDateRef = useRef<HTMLButtonElement>(null);
    useFocusTrap(todayOrSelectedDateRef.current && wrapperRef.current, todayOrSelectedDateRef.current);

    const onTextFieldChange = (textFieldValue: string, textFieldName?: string, event?: SyntheticEvent) => {
        if (!textFieldValue) {
            onChange(undefined, textFieldName, event);
        }
    };

    const onDatePickerChange = (newDate?: Date) => {
        onChange(newDate, name);
        onClose();
    };

    return (
        <>
            <TextField
                ref={ref}
                {...forwardedProps}
                name={name}
                forceFocusStyle={isOpen}
                textFieldRef={anchorRef}
                value={value ? moment(value).locale(locale).format('LL') : ''}
                onClick={toggleSimpleMenu}
                onChange={onTextFieldChange}
                onKeyPress={handleKeyboardNav}
                isDisabled={isDisabled}
                readOnly
            />
            {isOpen ? (
                <Popover
                    anchorRef={anchorRef}
                    placement={Placement.BOTTOM_START}
                    isOpen={isOpen}
                    onClose={onClose}
                    closeOnClickAway
                    closeOnEscape
                >
                    <div ref={wrapperRef}>
                        <DatePicker
                            locale={locale}
                            maxDate={maxDate}
                            minDate={minDate}
                            value={value}
                            onChange={onDatePickerChange}
                            todayOrSelectedDateRef={todayOrSelectedDateRef}
                            defaultMonth={defaultMonth}
                            nextButtonProps={nextButtonProps}
                            previousButtonProps={previousButtonProps}
                        />
                    </div>
                </Popover>
            ) : null}
        </>
    );
});
DatePickerField.displayName = COMPONENT_NAME;
