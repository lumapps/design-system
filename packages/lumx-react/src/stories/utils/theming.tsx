import React from 'react';
import { Predicate } from '@lumx/react/utils/type';
import { FlexBox } from '@lumx/react';

type Property = string | { value: string };
interface PropertyTree {
    [key: string]: PropertyTree | Property;
}
type Entry = [key: string, value: string];
type Variables = Record<string, string>;

function asValue(prop?: PropertyTree | Property): string | undefined {
    if (typeof prop === 'string') {
        return prop;
    }
    if (typeof prop === 'object' && 'value' in prop) {
        return prop.value as string;
    }
    return undefined;
}

/** Convert property tree into a list of args entries ([key, value]) */
function* toArgEntries(props: PropertyTree | Property, parent = ''): Iterable<Entry> {
    const value = asValue(props);
    if (value) {
        yield [parent, value];
    } else {
        for (const [key, prop] of Object.entries(props)) {
            const newParent = parent ? `${parent}-${key}` : key;
            yield* toArgEntries(prop, newParent);
        }
    }
}
function* parseVariableEntries(cssString: string, strict = false) {
    for (const line of cssString.split(/(?<=;)/)) {
        if (!line.trim()) continue;
        const match = line.match(/^\s*(--lumx[^:]+)\s*:\s*([^;]+);/m);
        if (!match) {
            if (strict) throw new Error(`Could not parse \`${line}\``);
            else continue;
        }
        const [, variableName, variableValue] = match;
        yield [variableName, variableValue];
    }
}

function VariableEditor({
    variable: [key, value],
    setVariable,
    defaultStyle,
}: {
    variable: Entry;
    setVariable: (newVariable: Entry) => void;
    defaultStyle: Variables;
}) {
    const onChange = React.useCallback<React.ChangeEventHandler<HTMLInputElement>>(
        (evt) => {
            setVariable([evt.target.name, evt.target.value]);
        },
        [setVariable],
    );

    return (
        <FlexBox as="label" orientation="horizontal" hAlign="center" style={{ whiteSpace: 'nowrap' }}>
            <span style={{ flexShrink: 0 }}>{key}</span>:
            <input
                name={key}
                onChange={onChange}
                style={{ border: 0, padding: 8, fieldSizing: 'content', display: 'inline' } as any}
                value={defaultStyle[key] === value ? '' : value}
                placeholder={value}
            />
            ;
        </FlexBox>
    );
}

function VariablesEditor({
    style,
    defaultStyle,
    setStyle,
}: {
    style: Variables;
    defaultStyle: Variables;
    setStyle: (fn: (newStyle: Variables) => Variables) => void;
}) {
    const setVariable = React.useCallback(
        ([key, value]: Entry) => {
            setStyle((prevStyle) => ({ ...prevStyle, [key]: value }));
        },
        [setStyle],
    );
    return (
        <details style={{ fontFamily: 'monospace' }} className="lumx-spacing-margin-vertical-huge">
            <summary className="lumx-typography-subtitle2">Edit variables</summary>
            {Object.entries(style).map((variable) => (
                <VariableEditor
                    key={variable[0]}
                    defaultStyle={defaultStyle}
                    variable={variable}
                    setVariable={setVariable}
                />
            ))}
        </details>
    );
}

const FILTER_PROPS = ['theme', 'emphasis', 'size'];

export function withTheming<P extends PropertyTree>({
    properties,
    values,
}: {
    properties?: P;
    values?: string | Partial<P>;
}) {
    const initialStyle = values
        ? Object.fromEntries(
              typeof values === 'string'
                  ? parseVariableEntries(values)
                  : toArgEntries(values as PropertyTree, '--lumx'),
          )
        : [];
    const defaultStyle = Object.fromEntries(properties ? Array.from(toArgEntries(properties, '--lumx')) : []);

    const varNames = new Set(Object.keys(initialStyle).concat(Object.keys(defaultStyle)));

    function filteredStyles(args: any) {
        // Create predicate
        const filterArgs = Object.entries(args)
            .map(([argName, argValue]) => {
                if (typeof argValue === 'string' && FILTER_PROPS.includes(argName)) {
                    return (varName: string) =>
                        // Has the given arg with value
                        varName.match(new RegExp(`${argName}-${argValue}`)) ||
                        (argName === 'emphasis' && argValue === 'medium' && varName.match(/emphasis-selected/)) ||
                        // Or does not have the arg at all
                        !varName.match(new RegExp(`${argName}-\\w+`));
                }
                return false;
            })
            .filter(Boolean) as Array<Predicate<string>>;
        const styles = Object.fromEntries(
            Array.from(varNames.values())
                .map((varName) => {
                    if (!filterArgs.every((filter) => filter(varName))) return null as any;
                    return [varName, initialStyle[varName] || defaultStyle[varName]];
                })
                .filter(Boolean),
        );
        return () => styles;
    }

    return (Story: any, ctx: any) => {
        const [style, setStyle] = React.useState<Variables>(filteredStyles(ctx.args));

        return (
            <>
                <Story args={{ ...ctx.args, style: { ...ctx.args.style, ...style } }} />
                <VariablesEditor style={style} defaultStyle={defaultStyle} setStyle={setStyle} />
            </>
        );
    };
}
