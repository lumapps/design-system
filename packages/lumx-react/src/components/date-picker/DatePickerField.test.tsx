import React, { ReactElement } from 'react';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';

import { Wrapper } from '@lumx/react/testing/utils';

import { DatePickerField, DatePickerFieldProps } from './DatePickerField';

const mockedDate = new Date(
    new Date(1487721600).toLocaleString('en-US', {
        timeZone: 'America/Toronto',
    }),
);
Date.now = jest.fn(() => mockedDate.valueOf());

type SetupProps = Partial<DatePickerFieldProps>;

const setup = ({ ...propsOverride }: SetupProps = {}, shallowRendering = true) => {
    const props: DatePickerFieldProps = {
        label: 'DatePickerField',
        locale: 'fr',
        onChange: jest.fn(),
        value: mockedDate,
        nextButtonProps: { label: 'Next month' },
        previousButtonProps: { label: 'Previous month' },
        ...propsOverride,
    };
    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;
    const wrapper: Wrapper = renderer(<DatePickerField {...props} />);

    return { props, wrapper };
};

describe(`<${DatePickerField.displayName}>`, () => {
    describe('Snapshots and structure', () => {
        it('should render correctly', () => {
            const { wrapper } = setup();
            expect(wrapper).toMatchSnapshot();
        });

        it('should render correctly when passed a date object', () => {
            const { wrapper } = setup({ value: new Date('January 18, 1970') });
            expect(wrapper).toMatchSnapshot();
        });
    });
});
