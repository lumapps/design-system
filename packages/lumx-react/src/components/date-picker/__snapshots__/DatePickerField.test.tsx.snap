// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePickerField> Snapshots and structure should render correctly 1`] = `
<Fragment>
  <TextField
    forceFocusStyle={false}
    label="DatePickerField"
    onChange={[Function]}
    onClick={[Function]}
    onKeyPress={[Function]}
    readOnly={true}
    textFieldRef={
      Object {
        "current": null,
      }
    }
    theme="light"
    type="text"
    value="18 janvier 1970"
  />
</Fragment>
`;

exports[`<DatePickerField> Snapshots and structure should render correctly when passed a date object 1`] = `
<Fragment>
  <TextField
    forceFocusStyle={false}
    label="DatePickerField"
    onChange={[Function]}
    onClick={[Function]}
    onKeyPress={[Function]}
    readOnly={true}
    textFieldRef={
      Object {
        "current": null,
      }
    }
    theme="light"
    type="text"
    value="18 janvier 1970"
  />
</Fragment>
`;
